### ISSUE TITLE
API for status line updates

### ISSUE THREAD

**Issue by nweisenfeld** [napari/napari#2253]
## 🚀 Feature

The bottom line of the viewer, which I'm calling the "status line," has a display that changes with cursor position and active layer.  The goal here is to make the display more configurable, in order to display application-specific information about the location of the cursor.  For instance, displaying label names e.g.  ("cancer", "not cancer"), instead of or in addition to a numeric index.

## Motivation

For an RGB image layer, you see something like:

layer name [1951 542]: [128, 255, 0]

where the first set of numbers is the coordinate and the second set are RGB (BGR?) values.

For a points layer, the display shows:

layer name [1951 542]
layer name [1980 5151]: 1234

indicating coordinates only when the cursor is not over a point and adding the point _index_, when the cursor is over a point.

More useful in this latter case would be the ability to customize the display to show metadata or properties associated with the point.

## Pitch

I'm too new to Napari to have a strong opinion about the best approach, but here are some ideas:

We could take a  granular approach where layer-specific options are enabled.  For instance, a "status_properties" parameter to a points layer where the properties are required to either be arrays of the same length as the data or a dictionary from point index to property.

Or we could take a more general approach where the status line generation in simply overridable in some clean way, possibly by making  [`generate_layer_status()`](https://github.com/napari/napari/blob/154f15f96fcef795f8ddbe5bb4b981ec5bf2ce18/napari/utils/status_messages.py#L43) a function within the layer class and user-overridable.

Or combine the two - have layer-specific defaults in `generate_layer_status`, but have it overridable.

The latter exists already, but in a slight abuse of the current structure.  It's possible to override [`layer.get_value()`](https://github.com/napari/napari/blob/154f15f96fcef795f8ddbe5bb4b981ec5bf2ce18/napari/layers/base/base.py#L888), since this is what is ultimately passed to [`generate_layer_status()`](https://github.com/napari/napari/blob/154f15f96fcef795f8ddbe5bb4b981ec5bf2ce18/napari/utils/status_messages.py#L43) to generate the status line.

## Alternatives

As mentioned in the last section, this can be done already, to a degree, by overriding `get_value()` for a layer.  PR #2235 updates get_value() to behave properly when confronted with a string, so an arbitrary string could be used here.  But that's clearly an abuse of get_value() and could have repercussions elsewhere.  It's likely not the case that we'd want the "value" in a computational sense to match the user friendly display.

## Additional context

cc: @VolkerH

**Comment by sofroniewn** [comment]
Hi @nweisenfeld thanks for raising this as an issue after your PR earlier. I'm wondering the best way to do this.

> We could take a granular approach where layer-specific options are enabled. For instance, a "status_properties" parameter to a points layer where the properties are required to either be arrays of the same length as the data or a dictionary from point index to property.

The granular approach is appealing as I could imagine options in a settings and preferences panel that would allow a GUI only user to customize what they see in the status bar - "I only want to see coordinates", "I only want to see labels" etc.

That might be too limiting though for all the things someone wants to do though, so maybe letting someone pass a custom function as an option too might be desired.

I know @VolkerH has been quite interested in improvements here for a while too so thanks for mentioning him in this issue. @VolkerH do you have further ideas on what you'd want to see from an API here?

Let's keep discussing

**Comment by VolkerH** [comment]
Hi,
thanks for tagging me.

Well, I had the specific use case of displaying the zoom level in mind. I thought that should go into the status bar (it is a very common feature). 

I am not sure whether I would want the general functionality to completely change the status bar. A plugin could then completely change the user experience.

For additional information, maybe a docked inspector widget that is customizable would be the way to go? 

I did something hacky a long while back (early days of napari) to explore measurement values: 

https://raw.githubusercontent.com/VolkerH/NapariPlayground/master/Gallery/napari_label_inspector.gif

For the general use case of showing some metadata of the point or object where the mouse cursor is, maybe such a separate widget (obviouly more nicely integrated than the hack above) would be an option. This would keep the status bar for essentials that the user can rely on to be there. (It will be hard to decide what is essential. For me, the zoom level is high on the list, but others may have different priorities)

**Comment by sofroniewn** [comment]
> For additional information, maybe a docked inspector widget that is customizable would be the way to go?

Yes I like this idea too. I do worry that the status bar might get a little long, cluttered otherwise. 

> This would keep the status bar for essentials that the user can rely on to be there. (It will be hard to decide what is essential. For me, the zoom level is high on the list, but others may have different priorities)

This is something that could be configurable, we could provide a list of 5 or so different things people might want to see and the people could check or uncheck the exact ones

**Comment by goanpeca** [comment]
@sofroniewn will there be some sort of plugin API to hook an add a "status bar widget" that provides additional features?

**Comment by sofroniewn** [comment]
> @sofroniewn will there be some sort of plugin API to hook an add a "status bar widget" that provides additional features?

Could be, i'd be very open to a hook spec that allowed people to update the status bar. Would have to thing through the details, but in theory i like it. Do you have something in mind @goanpeca? Curious @tlambert03 if you have any ideas here too?

**Comment by goanpeca** [comment]
Nothing in particular, but at some point someone would probably want to add a widget that displays something that will make sense for a specific plugin, and cannot be conveyed via the normal status bar text. But this is probably a topic that deserves a separate issue.

**Comment by svikolev** [comment]
Would like the ability to se the status bar to display the value of more then 1 layer. currently displays position and value of selected layer if it is visible. I would like to display the value of other layers as well. Not all of them but the ones I set.
